poc/vidpf.py

```diff
@@ -11,275 +10,365 @@
-
-    def eval_next(self, prev_seed, prev_ctrl, correction_word, cs_proof,
-                  current_level, node, pi_proof, binder):
+            w = prefix_tree_share[(prefix, level)][2]
+            out_share.append(w if agg_id == 0 else vec_neg(w))
+
+        # Compute the evaluation proof. If both aggregators compute the
+        # same value, then they agree on the walk proof, path and
+        # counter.
+        proof = eval_proof(proof, counter, path)
+        return (beta_share, out_share, proof)
+
+    def eval_next(self,
+                  seed: bytes,
+                  ctrl: Field2,
+                  correction_word: CorrectionWord,
+                  i: int,  # current level
+                  node: int,
+                  proof: bytes,
+                  nonce: bytes,
+                  ) -> tuple[bytes, Field2, list[F], bytes]:
         """
-        Compute the next node in the VIDPF tree along the path determined by
-        a candidate prefix. The next node is determined by `bit`, the bit of
-        the prefix corresponding to the next level of the tree.
+        Extend a node in the tree, select and correct one of its
+        children, then convert it into a payload and the next seed.
         """
-        (seed_cw, ctrl_cw, w_cw) = correction_word
-
-        # (s^L, s^R), (t^L, t^R) = PRG(s^{i-1})
-        (s, t) = self.extend(prev_seed, binder)
-        s[0] = xor(s[0], prev_ctrl.conditional_select(seed_cw))  # s^L
-        s[1] = xor(s[1], prev_ctrl.conditional_select(seed_cw))  # s^R
-        t[0] += ctrl_cw[0] * prev_ctrl  # t^L
-        t[1] += ctrl_cw[1] * prev_ctrl  # t^R
-
-        bit = node & 1
-        next_ctrl = t[bit]  # t'^i
-        (next_seed, w) = self.convert(s[bit],  binder)  # s^i, W^i
-        # Implementation note: Here we add the correction word to the
-        # output if `next_ctrl` is set. We avoid branching on the value of
-        # the control bit in order to reduce side channel leakage.
-        y = []
-        mask = self.field(next_ctrl.as_unsigned())
-        for i in range(len(w)):
-            y.append(w[i] + w_cw[i] * mask)
-
-        # pi' = H(x^{<= i} || s^i)
-        pi_prime = self.next_cs_proof(node, current_level, next_seed)
-
-        # \pi = \pi xor H(\pi \xor (proof_prime \xor next_ctrl * cs_proof))
-        if next_ctrl.as_unsigned() == 1:
-            h2 = xor(pi_proof, xor(pi_prime, cs_proof))
+        (seed_cw, ctrl_cw, w_cw, proof_cw) = correction_word
+        keep = node & 1
+
+        # Extend.
+        #
+        # [MST24, Fig. 17]: (s^L, s^R), (t^L, t^R) = PRG(s^{i-1})
+        (s, t) = self.extend(seed, nonce)
+
+        # Correct.
+        #
+        # Implementation note: avoid branching on the value of control bits, as
+        # its value may be leaked by a side channel.
+        if ctrl == Field2(1):
+            s[keep] = xor(s[keep], seed_cw)
+            t[keep] = t[keep] + ctrl_cw[keep]
```

Comment:
What we had before here avoided branching:
```
s[0] = xor(s[0], prev_ctrl.conditional_select(seed_cw))  # s^L
s[1] = xor(s[1], prev_ctrl.conditional_select(seed_cw))  # s^R
t[0] += ctrl_cw[0] * prev_ctrl  # t^L
t[1] += ctrl_cw[1] * prev_ctrl  # t^R
```
What you added is cleaner but introduces branching. Is there any other benefit we get?

Reply:
Clearer is better I think: I'm trying to make this code as readable as possible. It's up to implementations to make things constant time.